### PR TITLE
[OG-ATTACHMENTS-5] PLU-769 (feat): multi-attachment updates with append/replace in update-case

### PR DESCRIPTION
**Pull request by kevinkim-ogp**
## TL;DR
The Ownself Gather **Update case** action can now upload file attachments into one or more of a case's attachment fields, with a per-field choice to **append** to (default) or **replace** the field's existing attachments.

## What changed?
- New action input: **Attachment fields**, a multirow list where each row picks an attachment field, an **Update mode** (Add to existing / Replace existing, default add), and the files to upload.
- On run: for each row, uploads the files, then either replaces the field outright or merges them with the field's existing uuids (de-duplicated) so unrelated existing attachments are preserved.
- Validation: a selected field needs at least one file, and the same attachment field cannot be selected twice. Clearing the field ignores leftover hidden files instead of failing the step.
- 422 `INVALID_INPUT` now also tells users to stay within the attachment maximum configured in Ownself Gather.
- Fixed the "tested" success state not showing after a successful test run: the frontend's dataIn-vs-parameters comparison only special-cased a top-level `attachments` key, so it never matched the nested `attachmentFields[i].attachments` shape and always reported a mismatch.
- Action and dynamic-data tests extended.

## Tests (manual)
- [ ] Add the Ownself Gather "Update case" action to a flow and connect with an API key.
- [ ] Enter a case UUID whose case type has attachment fields; confirm the **Attachment fields** row's **Field** dropdown lists them.
- [ ] Add a row, pick a field, leave **Update mode** at default (Add to existing), add a file, and test the step.
- [ ] In Gather, confirm the file is attached under that field, and any pre-existing attachments remain.
- [ ] Add another row for a second attachment field, set **Update mode** to "Replace existing attachments", test, and confirm Gather shows only the newly uploaded file(s) for that field.
- [ ] Add two rows pointing at the same field → confirm a clear validation error.
- [ ] Set a row's field but no files → confirm a clear validation error.
- [ ] Select files, then clear the field → the step still runs and leftover files are not uploaded.
- [ ] After a successful test, confirm the step shows the tested/success state and it persists on reopening the step.
- [ ] Change a step parameter after testing → confirm the tested state correctly clears.
- [ ] Regression: update normal text/number/null case fields with no attachments → unchanged behavior.
- [ ] Regression: update case status only → unchanged behavior.

## Deploy notes
No migrations or config changes. We are not pre-checking per-field `maxCount`. If Gather rejects the write, the 422 message now points at Ownself Gather's attachment maximum.

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds multi-field attachment uploads to Gather's Update case action, supporting append and replacement modes, and updates frontend test-state comparison for nested attachment inputs.

- Adds attachment-field configuration, validation, upload handling, and dynamic field selection.
- Reads and merges existing UUIDs when append mode is selected.
- Extends action tests for attachment payloads and validation.
- Extends frontend parameter comparison to recognize nested attachment arrays.
</details>

<details><summary><h3>Confidence Score: 3/5</h3></summary>

The PR should not merge until concurrent append updates cannot lose attachments and conflicting field inputs are rejected instead of silently overwritten.

Append mode replaces an array derived from a separate unguarded read, allowing concurrent runs to lose successful updates, while duplicate names across the two field inputs cause one configured value to be silently discarded.

**Files Needing Attention:** packages/backend/src/apps/gathersg/actions/update-case/index.ts, packages/backend/src/apps/gathersg/actions/update-case/schema.ts
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/backend/src/apps/gathersg/actions/update-case/index.ts | Adds attachment uploads and append/replace payload construction, but append is a non-atomic read-modify-write and cross-list field collisions silently overwrite values. |
| packages/backend/src/apps/gathersg/actions/update-case/schema.ts | Validates S3 IDs, paired field/file inputs, and duplicates within attachment rows, but does not reject names duplicated across ordinary and attachment fields. |
| packages/frontend/src/components/FlowStepTestController/utils.tsx | Adds nested attachment matching, with a non-blocking filename-only equality weakness for distinct objects sharing a filename. |
| packages/backend/src/apps/gathersg/__tests__/actions/update-case.test.ts | Covers primary attachment modes and validation but omits concurrent append and cross-list field-collision cases. |
| packages/backend/src/apps/gathersg/common/constants.ts | Introduces the attachment field-type constant used by upload-token requests. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
  participant Flow as Plumber flow
  participant Action as Update case action
  participant FileAPI as Gather File API
  participant CaseAPI as Gather Case API

  Flow->>Action: Run with attachment field rows
  loop Each attachment field
    Action->>FileAPI: Upload selected files
    FileAPI-->>Action: Uploaded file UUIDs
    alt Add to existing
      Action->>CaseAPI: GET case
      CaseAPI-->>Action: Existing attachment UUIDs
      Action->>Action: Merge and deduplicate UUIDs
    else Replace existing
      Action->>Action: Use uploaded UUIDs only
    end
  end
  Action->>CaseAPI: PATCH case with complete field arrays
  CaseAPI-->>Action: Trace ID
```
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Greploop%20opengovsg%2Fplumber%20PR%20%231714%3A%20work%20through%20Greptile's%20open%20review%20comments%2C%20then%20keep%20reviewing%20and%20fixing%20until%20it%20comes%20back%20clean%20at%205%2F5%20with%20zero%20unresolved%20comments.%0AStart%20by%20reading%20the%20comments%20off%20the%20PR%20itself.%20On%20GitHub%2C%20use%20paginated%20%60gh%20api%20graphql%60%20to%20query%20%60pullRequest.reviewThreads%60%20with%20each%20thread's%20%60isResolved%60%20value%20and%20inline%20%60comments%60.%20%60gh%20pr%20view%20--comments%60%20only%20includes%20conversation%20comments%2C%20so%20do%20not%20use%20it%20as%20the%20findings%20source.%20They%20are%20not%20listed%20in%20this%20prompt%20on%20purpose%3A%20the%20PR%20is%20current%2C%20a%20pasted%20copy%20would%20not%20be.%20Skip%20anything%20already%20resolved%2C%20and%20if%20you%20judge%20a%20comment%20wrong%2C%20say%20so%20rather%20than%20changing%20code%20to%20satisfy%20it.%0A%0APrefer%20the%20Greptile%20CLI%2C%20which%20reviews%20the%20working%20tree%20with%20no%20push%20and%20no%20CI%20run.%20Fall%20back%20to%20PUSH%20LOOP%20only%20where%20a%20step%20below%20says%20to.%0A1.%20Run%20%60command%20-v%20greptile%60.%20Missing%3A%20go%20to%20PUSH%20LOOP%2C%20and%20pitch%20the%20CLI%20at%20the%20end.%0A2.%20Run%20%60greptile%20whoami%60.%20It%20exits%200%20even%20when%20signed%20out%2C%20so%20read%20the%20output%3A%20if%20it%20says%20%22Not%20signed%20in%22%2C%20ask%20the%20user%20to%20run%20%60greptile%20login%60%20%28it%20opens%20a%20browser%2C%20do%20not%20run%20it%20yourself%29%20and%20go%20to%20PUSH%20LOOP%20with%20no%20pitch.%0A3.%20Look%20for%20%60greploop%2FSKILL.md%60%20under%20%60.agents%2Fskills%2F%60%20or%20%60.claude%2Fskills%2F%60%2C%20both%20in%20the%20repo%20root%20and%20in%20the%20user's%20home%20directory.%20Found%3A%20follow%20that%20skill%2C%20never%20pushing%20between%20rounds%2C%20and%20you%20are%20done.%0A4.%20Missing%3A%20run%20%60greptile%20skills%20list%60.%20If%20it%20fails%20as%20an%20unknown%20command%2C%20this%20CLI%20predates%20the%20skill%2C%20so%20tell%20the%20user%20to%20run%20%60greptile%20update%60%20and%20go%20to%20PUSH%20LOOP%20with%20no%20pitch.%20Otherwise%20ask%20the%20user%20before%20running%20%60greptile%20skills%20install%20greploop%60%2C%20then%20follow%20the%20skill%20it%20writes.%0A%0APUSH%20LOOP%20%28only%20when%20a%20step%20above%20sends%20you%20here%29%3A%0A%0AGREPLOOP%20ENABLED%3A%20After%20fixing%20all%20issues%2C%20commit%20and%20push%20your%20changes.%20Greptile%20will%20automatically%20re-review%20the%20PR%20via%20webhook.%20Wait%20for%20the%20review%20to%20complete%2C%20then%20read%20the%20new%20review%20comments%20and%20confidence%20score.%20If%20the%20confidence%20score%20is%20below%205%2F5%20or%20there%20are%20unresolved%20comments%2C%20fix%20those%20issues%20and%20push%20again.%20Repeat%20until%20the%20confidence%20score%20is%205%2F5%20with%20zero%20unresolved%20comments.%20Maximum%205%20iterations.%20Do%20NOT%20open%20a%20new%20PR%20%E2%80%94%20keep%20pushing%20to%20the%20same%20branch.%0A%0AIF%20YOU%20ENDED%20UP%20ON%20PUSH%20LOOP%20BECAUSE%20THE%20CLI%20WAS%20MISSING%3A%20once%20the%20loop%20is%20done%2C%20print%20the%20block%20below%20and%20nothing%20else%20about%20the%20CLI.%20Reproduce%20it%20character%20for%20character%20inside%20a%20fenced%20code%20block%2C%20so%20the%20borders%20line%20up.%20Do%20not%20reword%20it%2C%20translate%20it%2C%20summarize%20it%2C%20add%20to%20it%2C%20or%20drop%20the%20box.%20Print%20it%20once%2C%20and%20do%20not%20bring%20the%20CLI%20up%20again%20later%20in%20the%20session.%0A%0A%2B----------------------------------------------------------------------------%2B%0A%7C%20%20You%20can%20run%20greploops%20faster%20locally%20with%20our%20CLI.%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7C%0A%7C%20%20Install%20it%20at%20https%3A%2F%2Fwww.greptile.com%2Fcli%2C%20or%20I%20can%20install%20it%20for%20you.%20%20%7C%0A%2B----------------------------------------------------------------------------%2B%0A%0AIf%20they%20take%20you%20up%20on%20it%2C%20install%20with%20%60npm%20install%20-g%20greptile%60%20%28or%20%60brew%20install%20greptileai%2Ftap%2Fgreptile%60%29%2C%20then%20%60greptile%20skills%20install%20greploop%60.%20Leave%20%60greptile%20login%60%20to%20them%2C%20it%20opens%20a%20browser.&repo=opengovsg%2Fplumber&pr=1714&platform=github">Fix all with Greploop</a> <a href="https://app.greptile.com/ide/claude-code?prompt=%23%23%23%20Issue%201%0Apackages%2Fbackend%2Fsrc%2Fapps%2Fgathersg%2Factions%2Fupdate-case%2Findex.ts%3A244-255%0A**Append%20updates%20lose%20attachments**%0A%0AIf%20two%20executions%20append%20to%20the%20same%20case%20field%20concurrently%2C%20each%20derives%20a%20complete%20replacement%20array%20from%20its%20own%20earlier%20GET%20and%20PATCHes%20it%20without%20a%20version%20check%20or%20case-scoped%20lock%2C%20causing%20the%20later%20write%20to%20remove%20attachments%20added%20by%20the%20earlier%20execution.%0A%0A%23%23%23%20Issue%202%0Apackages%2Fbackend%2Fsrc%2Fapps%2Fgathersg%2Factions%2Fupdate-case%2Findex.ts%3A257-260%0A**Conflicting%20fields%20overwrite%20silently**%0A%0AWhen%20a%20user%20enters%20an%20attachment%20field%20name%20in%20the%20arbitrary%20Case%20fields%20input%20and%20selects%20the%20same%20field%20under%20Attachment%20fields%2C%20this%20assignment%20silently%20replaces%20the%20configured%20case-field%20value%20with%20the%20UUID%20array%2C%20causing%20the%20submitted%20update%20to%20differ%20from%20the%20configured%20inputs.%0A%0A%23%23%23%20Issue%203%0Apackages%2Ffrontend%2Fsrc%2Fcomponents%2FFlowStepTestController%2Futils.tsx%3A107-111%0A**Filename%20matching%20hides%20changes**%0A%0AThe%20nested%20attachment%20comparison%20discards%20the%20S3%20path%20and%20compares%20only%20the%20final%20filename%2C%20so%20replacing%20an%20attachment%20with%20a%20different%20object%20that%20has%20the%20same%20filename%20leaves%20the%20successful-test%20state%20marked%20as%20current.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=opengovsg%2Fplumber&pr=1714&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=%23%23%23%20Issue%201%0Apackages%2Fbackend%2Fsrc%2Fapps%2Fgathersg%2Factions%2Fupdate-case%2Findex.ts%3A244-255%0A**Append%20updates%20lose%20attachments**%0A%0AIf%20two%20executions%20append%20to%20the%20same%20case%20field%20concurrently%2C%20each%20derives%20a%20complete%20replacement%20array%20from%20its%20own%20earlier%20GET%20and%20PATCHes%20it%20without%20a%20version%20check%20or%20case-scoped%20lock%2C%20causing%20the%20later%20write%20to%20remove%20attachments%20added%20by%20the%20earlier%20execution.%0A%0A%23%23%23%20Issue%202%0Apackages%2Fbackend%2Fsrc%2Fapps%2Fgathersg%2Factions%2Fupdate-case%2Findex.ts%3A257-260%0A**Conflicting%20fields%20overwrite%20silently**%0A%0AWhen%20a%20user%20enters%20an%20attachment%20field%20name%20in%20the%20arbitrary%20Case%20fields%20input%20and%20selects%20the%20same%20field%20under%20Attachment%20fields%2C%20this%20assignment%20silently%20replaces%20the%20configured%20case-field%20value%20with%20the%20UUID%20array%2C%20causing%20the%20submitted%20update%20to%20differ%20from%20the%20configured%20inputs.%0A%0A%23%23%23%20Issue%203%0Apackages%2Ffrontend%2Fsrc%2Fcomponents%2FFlowStepTestController%2Futils.tsx%3A107-111%0A**Filename%20matching%20hides%20changes**%0A%0AThe%20nested%20attachment%20comparison%20discards%20the%20S3%20path%20and%20compares%20only%20the%20final%20filename%2C%20so%20replacing%20an%20attachment%20with%20a%20different%20object%20that%20has%20the%20same%20filename%20leaves%20the%20successful-test%20state%20marked%20as%20current.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&pr=1714&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=6">Fix All in Cursor</picture></a> <a href="https://app.greptile.com/api/ide/conductor?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22opengovsg%2Fplumber%22%20on%20the%20existing%20branch%20%22og-attachments%2Fupdate-case-upload%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22og-attachments%2Fupdate-case-upload%22.%0A%0A%23%23%23%20Issue%201%0Apackages%2Fbackend%2Fsrc%2Fapps%2Fgathersg%2Factions%2Fupdate-case%2Findex.ts%3A244-255%0A**Append%20updates%20lose%20attachments**%0A%0AIf%20two%20executions%20append%20to%20the%20same%20case%20field%20concurrently%2C%20each%20derives%20a%20complete%20replacement%20array%20from%20its%20own%20earlier%20GET%20and%20PATCHes%20it%20without%20a%20version%20check%20or%20case-scoped%20lock%2C%20causing%20the%20later%20write%20to%20remove%20attachments%20added%20by%20the%20earlier%20execution.%0A%0A%23%23%23%20Issue%202%0Apackages%2Fbackend%2Fsrc%2Fapps%2Fgathersg%2Factions%2Fupdate-case%2Findex.ts%3A257-260%0A**Conflicting%20fields%20overwrite%20silently**%0A%0AWhen%20a%20user%20enters%20an%20attachment%20field%20name%20in%20the%20arbitrary%20Case%20fields%20input%20and%20selects%20the%20same%20field%20under%20Attachment%20fields%2C%20this%20assignment%20silently%20replaces%20the%20configured%20case-field%20value%20with%20the%20UUID%20array%2C%20causing%20the%20submitted%20update%20to%20differ%20from%20the%20configured%20inputs.%0A%0A%23%23%23%20Issue%203%0Apackages%2Ffrontend%2Fsrc%2Fcomponents%2FFlowStepTestController%2Futils.tsx%3A107-111%0A**Filename%20matching%20hides%20changes**%0A%0AThe%20nested%20attachment%20comparison%20discards%20the%20S3%20path%20and%20compares%20only%20the%20final%20filename%2C%20so%20replacing%20an%20attachment%20with%20a%20different%20object%20that%20has%20the%20same%20filename%20leaves%20the%20successful-test%20state%20marked%20as%20current.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=opengovsg%2Fplumber"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInConductorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInConductor.svg?v=6">Fix All in Conductor</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
### Issue 1
packages/backend/src/apps/gathersg/actions/update-case/index.ts:244-255
**Append updates lose attachments**

If two executions append to the same case field concurrently, each derives a complete replacement array from its own earlier GET and PATCHes it without a version check or case-scoped lock, causing the later write to remove attachments added by the earlier execution.

### Issue 2
packages/backend/src/apps/gathersg/actions/update-case/index.ts:257-260
**Conflicting fields overwrite silently**

When a user enters an attachment field name in the arbitrary Case fields input and selects the same field under Attachment fields, this assignment silently replaces the configured case-field value with the UUID array, causing the submitted update to differ from the configured inputs.

### Issue 3
packages/frontend/src/components/FlowStepTestController/utils.tsx:107-111
**Filename matching hides changes**

The nested attachment comparison discards the S3 path and compares only the final filename, so replacing an attachment with a different object that has the same filename leaves the successful-test state marked as current.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(frontend): match nested attachment f..."](https://github.com/opengovsg/plumber/commit/469f58328166d2588391bfafb9faddec14d3ad9b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=59383061)</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

**Comment by kevinkim-ogp** [comment]
* **#2073** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/2073?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1715** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1715?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1714** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1714?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1714?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1713** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1713?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1712** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1712?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1673** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1673?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1672** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1672?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop-v2`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>lfg</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/opengovsg?ref=merge-queue-instructions-in-stack-comment&prId=1714).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts [hunk 1]
```diff
@@ -114,12 +141,74 @@ const action: IRawAction = {
...
+          patchBody.caseUuid,
+          attachmentField,
+        )
+        patchBody.fields = {
+          ...(patchBody.fields ?? {}),
+          [attachmentField]: [...new Set([...existingUuids, ...uuids])],
+        } as Record<string, string | number | null>
+      }
+
+      const rawResponse = await $.http.patch('/cases/:caseUuid', patchBody, {
```
Comment (m0nggh) [inline packages/backend/src/apps/gathersg/actions/update-case/index.ts:211]: find it quite weird that gathersg gives back 422 invalid field error when trying to upload another attachment when the max is 1... feels like this would be problematic

File: packages/backend/src/apps/gathersg/actions/update-case/schema.ts [hunk 2]
```diff
@@ -73,11 +75,54 @@ export const requestSchema = z
...
+        return result
+      })
+      .nullish(),
+  })
+  // attachmentField and attachments must be set together. Two directional
+  // refines give an accurate, attributable message for each missing side.
+  .refine((data) => !(data.attachments?.length && !data.attachmentField), {
+    message: 'Please select an attachment field for your attachments.',
+    path: ['attachmentField'],
+  })
```
Comment (m0nggh) [inline packages/backend/src/apps/gathersg/actions/update-case/schema.ts:115]: right now, if i un-select the attachment field, it will cause an error because of this: suggest to remove this refine and 

// and make the final transform ignore stale attachments (schema.ts:120-126):
```
.transform((data) => ({
  caseUuid: data.caseUuid,
  ...(data.caseStatus && { status: data.caseStatus }),
  fields: data.caseFields,
  attachmentField: data.attachmentField,
  // The UI hides (but does not clear) the attachments input when no field is
  // selected, so a cleared field can leave stale s3Ids behind. Ignore them
  // rather than failing the step.
  attachments: data.attachmentField ? (data.attachments ?? []) : [],
}))
```

Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/schema.ts:115]: should've thought of this earlier. can discuss on mon whether we want to add a field to get the user to say whether they are "adding" or "replacing" the attachments

**Review by m0nggh — DISMISSED** [review]
wanna check with you if we shoule have an option to let the users decide whether to replace/append attachments instead? i think this depends on the use cases out there right

File: packages/backend/src/apps/gathersg/actions/update-case/schema.ts (hunk 2 above)
Comment (m0nggh) [reply packages/backend/src/apps/gathersg/actions/update-case/schema.ts:115]: yessss HAHA i put in the comment omg naise lets discuss on mon ba

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts (hunk 1 above)
Comment (m0nggh) [reply packages/backend/src/apps/gathersg/actions/update-case/index.ts:211]: anw i realised we could get the `maxCount` in `fetchCaseFields` so we can technically check how many and then throw an error if someone tries to append more

**Comment by kevinkim-ogp** [comment]
thanks @m0nggh 🫡 
* add a radio for user to decide whether to append / replace attachments (default to append to err on the safe side so we don't accidentally replace attachments)
* add a check that the number of attachments the user selected in plumber does not exceed the limit
* improve error messaging if user exceeds the max number of attachments

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts (hunk 1 above)
Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/index.ts:304]: We'll monitor after release and assess whether this check is actually needed, rather than adding it upfront.

If we do add it, worth noting the overhead: `fetchCaseFields` needs `caseTypeUuid`, which requires an extra `GET /cases/:caseUuid` call first (we don't have it in hand when `replaceExisting` is true), plus the `GET /caseTypes/:caseTypeUuid` call itself. Both would run per attachment field row, per action execution, so cost scales with configured fields and flow execution volume rather than being a one-time config-time cost like the dynamic data dropdown.

File: packages/backend/src/apps/gathersg/actions/update-case/schema.ts [hunk 3]
```diff
@@ -1,8 +1,46 @@
...
+        context.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: `${value} is not a S3 ID.`,
+        })
+        return z.NEVER
+      }
+      result.push(value)
+    }
+    return result
+  })
```
[the remark this answers is not in the document]
Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/schema.ts:36]: !accepted_risk shared S3 helper, not unique to this action

`uploadCaseAttachments` already calls `getObjectFromS3Id` with `{ flowId }`, same as Postman. Bucket parsing and the mock-prefix metadata skip live in `s3.ts`. Fixing only this Zod schema would not close the path. Out of scope for this PR.

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts [hunk 4]
```diff
@@ -120,12 +154,114 @@ const action: IRawAction = {
...
+            : [
+                ...new Set([
+                  ...(await getExistingAttachmentUuids(
+                    $,
+                    patchBody.caseUuid,
+                    field,
+                  )),
+                  ...uuids,
+                ]),
+              ]
```
[the remark this answers is not in the document]
Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/index.ts:295]: Gather's PATCH replaces the full uuid array. Plumber cannot version or lock that write without Gather API support. Same last-write-wins behaviour as two people editing the case in Gather. Leaving as-is.

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts [hunk 5]
```diff
@@ -120,12 +154,114 @@ const action: IRawAction = {
...
+
+          patchBody.fields = {
+            ...(patchBody.fields ?? {}),
+            [field]: finalUuids,
+          } as Record<string, string | number | null>
```
[the remark this answers is not in the document]
Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/index.ts:300]: Attachment fields are already excluded from the Case fields dropdown (`UNSUPPORTED_FIELDS`). `allowArbitrary` can still let someone type the same name. In that case the attachment uuid array is the value Gather expects. Not adding extra collision validation.

File: packages/frontend/src/components/FlowStepTestController/utils.tsx [hunk 6]
```diff
@@ -86,6 +86,33 @@ const getTableData = (
...
+  }
+  return paramValue.every((attachment, index) => {
+    const lastTestAttachment = lastTest[index]
+    // manually uploaded attachments will have the same value
+    if (attachment === lastTestAttachment) {
+      return true
+    }
+
+    // attachments from FormSG will be using the s3 id
+    const lastTestFilename = String(lastTestAttachment).split('/').pop()
```
[the remark this answers is not in the document]
Comment (kevinkim-ogp) [reply packages/frontend/src/components/FlowStepTestController/utils.tsx:123]: This is the existing FormSG/Postman test-status rule. Step params store filenames; `dataIn` stores S3 paths, so matching has to use the filename. Same-name replacement leaving "test current" is an accepted edge case. Not changing it here.

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts (hunk 1 above)
Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/index.ts:304]: Not adding a per-field `maxCount` pre-check (extra case + case-type fetches on every run). 422 `INVALID_INPUT` now tells the user to keep attachment files at or below the maximum configured in Ownself Gather.

File: packages/backend/src/apps/gathersg/actions/update-case/index.ts [hunk 7]
```diff
@@ -162,11 +254,54 @@ const action: IRawAction = {
...
+            : [
+                ...new Set([
+                  ...(await getExistingAttachmentUuids(
+                    $,
+                    patchBody.caseUuid,
+                    field,
+                  )),
+                  ...uuids,
+                ]),
+              ]
```
[the remark this answers is not in the document]
Comment (kevinkim-ogp) [reply packages/backend/src/apps/gathersg/actions/update-case/index.ts:295]: !accepted_risk Gather PATCH is last-write-wins

Append is a GET then PATCH of the full uuid array because Gather has no append or versioned write. Two concurrent executions of the same case can drop each other's new files. That is the same race as two people editing the case in Gather. We are not adding a Plumber-side lock for this.

**Comment by ogp-weeloong** [comment]
Potentially a regression. I get a full page error:  
  
"**An error has occurred: "Path parameter caseUuid cannot be an empty string."**

If this error persists, please visit our [support form](https://go.gov.sg/plumber-support) for help."  
  
repro:

1. Create pipe that creates a case, then updates the case with UUID from the create case step
2. publish then unpublish
3. delete the create case step
4. try to open the "update case" step

**Review by ogp-weeloong — APPROVED** [review]
looks good! Got the full page error regression, can try rebasing

**Review by ogp-weeloong — APPROVED** [review]
